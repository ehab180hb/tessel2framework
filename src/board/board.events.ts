import * as events from 'events'
export const boardEvents = new events()

export enum BoardEvents {
  UPDATE_BUTTONS = 'updateButtons',
  UPDATE_SCREEN = 'updateScreen',
}
