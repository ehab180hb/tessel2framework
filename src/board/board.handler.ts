import * as five from 'johnny-five'

const isTesting = process.env.NODE_ENV === 'test'

const Tessel = !isTesting ? require('tessel-io') : class {}

export const getBoard = () => {
  return new five.Board({
    io: new Tessel(),
  })
}
