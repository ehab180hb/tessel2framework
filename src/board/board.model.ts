export enum PIN {
  'a0',
  'a1',
  'a2',
  'a3',
  'a4',
  'a5',
  'a6',
  'a7',
  'b0',
  'b1',
  'b2',
  'b3',
  'b4',
  'b5',
  'b6',
  'b7',
}
