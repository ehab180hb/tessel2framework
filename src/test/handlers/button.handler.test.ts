import * as five from 'johnny-five'

import {
  createdButtons,
  checkButtonExistence,
  findOrCreateButton,
  createButton,
  initDefaultButtons,
  findButton,
  clearButton,
  updateButton,
  updateButtonPress,
} from '../../button/button.handler'
import { PIN } from '../../board/board.model'
let buttonOnSpy: string[] = []
let buttonRemoveListenerSpy: string[] = []

beforeEach(() => {
  buttonOnSpy = []
  buttonRemoveListenerSpy = []
  ;(five.Button as any) = jest.fn(pin => ({
    on: (e: string, fn: Function) => {
      buttonOnSpy.push(JSON.stringify({ e, fn: fn.name, pin }))
    },
    removeAllListeners: (e?: string) => {
      buttonRemoveListenerSpy.push(JSON.stringify({ e, pin }))
    },
  }))
  ;(createdButtons as any) = [
    {
      name: 'red',
      pin: PIN.b5,
      listener: new five.Button(PIN.b5),
    },
    {
      name: 'green',
      pin: PIN.b6,
      listener: new five.Button(PIN.b6),
    },
    {
      name: 'blue',
      pin: PIN.b7,
      listener: new five.Button(PIN.b7),
    },
  ]
})

describe('Button handler tests', () => {
  describe('checkButtonExistence()', () => {
    test('When button exists', () => {
      const res = checkButtonExistence(PIN.b5, 'red')
      expect(res).toMatchSnapshot()
    })
    test('When button does not exit', () => {
      const res = checkButtonExistence(PIN.b1, 'white')
      expect(res).toMatchSnapshot()
    })
  })
  describe('findOrCreateButton()', () => {
    test('when button exists', () => {
      const res = findOrCreateButton(PIN.b5, 'red')
      expect(res).toMatchSnapshot()
    })
    test('when button does not exist', () => {
      const res = findOrCreateButton(PIN.b1, 'white')
      expect(res).toMatchSnapshot()
    })
  })
  describe('createButton()', () => {
    test('when pin and name are available', () => {
      const res = createButton(PIN.b2, 'white')
      expect(res).toMatchSnapshot()
    })
    test('when pin is occupied', async () => {
      expect(() => createButton(PIN.b5, 'white')).toThrowErrorMatchingSnapshot()
    })
    test('when name is taken', () => {
      expect(() => createButton(PIN.b1, 'red')).toThrowErrorMatchingSnapshot()
    })
  })
  describe('initDefaultButtons()', () => {
    test('when buttons are available', () => {
      const mockDefaultButtons = [
        {
          name: 'white',
          pin: PIN.b1,
        },
        {
          name: 'black',
          pin: PIN.b2,
        },
        {
          name: 'purple',
          pin: PIN.b3,
        },
      ]
      initDefaultButtons(mockDefaultButtons)
      expect(createdButtons).toMatchSnapshot()
    })
    test('when buttons are not available', () => {
      const mockDefaultButtons = [
        {
          name: 'red',
          pin: PIN.b5,
        },
        {
          name: 'green',
          pin: PIN.b6,
        },
        {
          name: 'blue',
          pin: PIN.b7,
        },
      ]
      initDefaultButtons(mockDefaultButtons)
      expect(createdButtons).toMatchSnapshot()
    })
  })
  describe('findButton()', () => {
    test('when button exists', () => {
      const res = findButton('red')
      expect(res).toMatchSnapshot()
    })
    test('when button does not exist', () => {
      const res = findButton('white')
      expect(res).toBeFalsy()
    })
  })
  describe('clearButton()', () => {
    test('when button exists', () => {
      clearButton('red')
      expect(buttonRemoveListenerSpy).toMatchSnapshot()
    })
    test('when button exists and event type specified', () => {
      clearButton('red', 'press')
      expect(buttonRemoveListenerSpy).toMatchSnapshot()
    })
    test('when button does not exist', () => {
      expect(() => clearButton('white')).toThrowErrorMatchingSnapshot()
    })
  })
  describe('updateButton()', () => {
    const cb = () => {}
    test('when button exists', () => {
      updateButton('red', 'hold', cb)
      expect({ buttonRemoveListenerSpy, buttonOnSpy }).toMatchSnapshot()
    })
    test('when button does not exist', () => {
      expect(() => updateButton('white', 'hold', cb)).toThrowErrorMatchingSnapshot()
    })
  })
  describe('updateButtonPress()', () => {
    const cb = () => {}
    test('when button exists', () => {
      updateButtonPress('red', cb)
      expect({ buttonRemoveListenerSpy, buttonOnSpy }).toMatchSnapshot()
    })
    test('when button does not exist', () => {
      expect(() => updateButtonPress('white', cb)).toThrowErrorMatchingSnapshot()
    })
  })
})
