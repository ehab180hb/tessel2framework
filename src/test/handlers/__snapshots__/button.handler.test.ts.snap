// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button handler tests checkButtonExistence() When button does not exit 1`] = `
Object {
  "nameTaken": undefined,
  "pinOccupied": undefined,
}
`;

exports[`Button handler tests checkButtonExistence() When button exists 1`] = `
Object {
  "nameTaken": Object {
    "listener": Object {
      "on": [Function],
      "removeAllListeners": [Function],
    },
    "name": "red",
    "pin": 13,
  },
  "pinOccupied": Object {
    "listener": Object {
      "on": [Function],
      "removeAllListeners": [Function],
    },
    "name": "red",
    "pin": 13,
  },
}
`;

exports[`Button handler tests clearButton() when button does not exist 1`] = `"Button white does not exist"`;

exports[`Button handler tests clearButton() when button exists 1`] = `
Array [
  "{\\"pin\\":13}",
]
`;

exports[`Button handler tests clearButton() when button exists and event type specified 1`] = `
Array [
  "{\\"e\\":\\"press\\",\\"pin\\":13}",
]
`;

exports[`Button handler tests createButton() when name is taken 1`] = `"Button name red is taken"`;

exports[`Button handler tests createButton() when pin and name are available 1`] = `
Object {
  "listener": Object {
    "on": [Function],
    "removeAllListeners": [Function],
  },
  "name": "white",
  "pin": 10,
}
`;

exports[`Button handler tests createButton() when pin is occupied 1`] = `"Pin 13 is occupied"`;

exports[`Button handler tests findButton() when button exists 1`] = `
Object {
  "listener": Object {
    "on": [Function],
    "removeAllListeners": [Function],
  },
  "name": "red",
  "pin": 13,
}
`;

exports[`Button handler tests findOrCreateButton() when button does not exist 1`] = `
Object {
  "listener": Object {
    "on": [Function],
    "removeAllListeners": [Function],
  },
  "name": "white",
  "pin": 9,
}
`;

exports[`Button handler tests findOrCreateButton() when button exists 1`] = `
Object {
  "listener": Object {
    "on": [Function],
    "removeAllListeners": [Function],
  },
  "name": "red",
  "pin": 13,
}
`;

exports[`Button handler tests initDefaultButtons() when buttons are available 1`] = `
Array [
  Object {
    "listener": Object {
      "on": [Function],
      "removeAllListeners": [Function],
    },
    "name": "red",
    "pin": 13,
  },
  Object {
    "listener": Object {
      "on": [Function],
      "removeAllListeners": [Function],
    },
    "name": "green",
    "pin": 14,
  },
  Object {
    "listener": Object {
      "on": [Function],
      "removeAllListeners": [Function],
    },
    "name": "blue",
    "pin": 15,
  },
  Object {
    "listener": Object {
      "on": [Function],
      "removeAllListeners": [Function],
    },
    "name": "white",
    "pin": 9,
  },
  Object {
    "listener": Object {
      "on": [Function],
      "removeAllListeners": [Function],
    },
    "name": "black",
    "pin": 10,
  },
  Object {
    "listener": Object {
      "on": [Function],
      "removeAllListeners": [Function],
    },
    "name": "purple",
    "pin": 11,
  },
]
`;

exports[`Button handler tests initDefaultButtons() when buttons are not available 1`] = `
Array [
  Object {
    "listener": Object {
      "on": [Function],
      "removeAllListeners": [Function],
    },
    "name": "red",
    "pin": 13,
  },
  Object {
    "listener": Object {
      "on": [Function],
      "removeAllListeners": [Function],
    },
    "name": "green",
    "pin": 14,
  },
  Object {
    "listener": Object {
      "on": [Function],
      "removeAllListeners": [Function],
    },
    "name": "blue",
    "pin": 15,
  },
]
`;

exports[`Button handler tests updateButton() when button does not exist 1`] = `"Button white does not exist"`;

exports[`Button handler tests updateButton() when button exists 1`] = `
Object {
  "buttonOnSpy": Array [
    "{\\"e\\":\\"hold\\",\\"fn\\":\\"cb\\",\\"pin\\":13}",
  ],
  "buttonRemoveListenerSpy": Array [
    "{\\"e\\":\\"hold\\",\\"pin\\":13}",
  ],
}
`;

exports[`Button handler tests updateButtonPress() when button does not exist 1`] = `"Button white does not exist"`;

exports[`Button handler tests updateButtonPress() when button exists 1`] = `
Object {
  "buttonOnSpy": Array [
    "{\\"e\\":\\"press\\",\\"fn\\":\\"cb\\",\\"pin\\":13}",
  ],
  "buttonRemoveListenerSpy": Array [
    "{\\"e\\":\\"press\\",\\"pin\\":13}",
  ],
}
`;
